wgpu-hal/src/dx12/mod.rs

```diff
@@ -65,39 +65,72 @@ use crate::auxil::{
     },
 };
 
+#[derive(Debug)]
+struct DynLib {
+    inner: libloading::Library,
+}
+
+impl DynLib {
+    unsafe fn new<P>(filename: P) -> Result<Self, libloading::Error>
+    where
+        P: AsRef<ffi::OsStr>,
+    {
+        unsafe { libloading::Library::new(filename) }.map(|inner| Self { inner })
+    }
+
+    unsafe fn get<T>(
+        &self,
+        symbol: &[u8],
+    ) -> Result<libloading::Symbol<'_, T>, crate::DeviceError> {
+        unsafe { self.inner.get(symbol) }.map_err(|e| match e {
+            libloading::Error::GetProcAddress { .. } | libloading::Error::GetProcAddressUnknown => {
+                crate::DeviceError::Unexpected
+            }
+            libloading::Error::IncompatibleSize
+            | libloading::Error::CreateCString { .. }
+            | libloading::Error::CreateCStringWithTrailing { .. } => crate::hal_internal_error(e),
+            _ => crate::DeviceError::Unexpected, // could be unreachable!() but we prefer to be more robust
+        })
+    }
+}
+
 #[derive(Debug)]
 struct D3D12Lib {
-    lib: libloading::Library,
+    lib: DynLib,
 }
 
 impl D3D12Lib {
     fn new() -> Result<Self, libloading::Error> {
-        unsafe { libloading::Library::new("d3d12.dll").map(|lib| D3D12Lib { lib }) }
+        unsafe { DynLib::new("d3d12.dll").map(|lib| Self { lib }) }
     }
 
     fn create_device(
         &self,
         adapter: &DxgiAdapter,
         feature_level: Direct3D::D3D_FEATURE_LEVEL,
-    ) -> Result<windows_core::Result<Direct3D12::ID3D12Device>, libloading::Error> {
+    ) -> Result<Direct3D12::ID3D12Device, crate::DeviceError> {
         // Calls windows::Win32::Graphics::Direct3D12::D3D12CreateDevice on d3d12.dll
         type Fun = extern "system" fn(
             padapter: *mut core::ffi::c_void,
             minimumfeaturelevel: Direct3D::D3D_FEATURE_LEVEL,
             riid: *const windows_core::GUID,
             ppdevice: *mut *mut core::ffi::c_void,
         ) -> windows_core::HRESULT;
-        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"D3D12CreateDevice") }?;
+        let func: libloading::Symbol<Fun> = unsafe { self.lib.get(b"D3D12CreateDevice\0") }?;
 
         let mut result__ = None;
-        Ok((func)(
-            unsafe { adapter.param().abi() },
+
+        (func)(
+            adapter.as_raw(),
             feature_level,
             // TODO: Generic?
             &Direct3D12::ID3D12Device::IID,
             <*mut _>::cast(&mut result__),
         )
-        .map(|| result__.expect("D3D12CreateDevice succeeded but result is NULL?")))
+        .ok()
+        .into_device_result("Device creation")?;
```

Comment:
This change is causing CI to panic (due to `device_lost_panic`). It seems that creating a device from the WARP adapter fails with `DXGI_ERROR_DEVICE_RESET` in the `DEVICE_LIFETIME_CHECK` & `MULTIPLE_DEVICES` tests.

Reply:
These 2 tests are not set up properly actually. They create a new instance, adapter and device from env or with defaults. I suspect that when we see they pass on WARP in CI, they actually use lavapipe.

Reply:
Does this mean we were never properly using WARP adapters? 👀

Reply:
I'm not sure, creating the first device from the WARP adapter seems to work but creating a 2nd one seems to fail with `DXGI_ERROR_DEVICE_RESET`.

Reply:
The only thing I could find online that is related to this is:

> Cheaper adapter enumeration
> For Direct3D 12, it’s no longer possible to backtrack from a device to the [IDXGIAdapter](https://learn.microsoft.com/en-us/windows/win32/api/DXGI/nn-dxgi-idxgiadapter) that was used to create it. It’s also no longer possible to provide D3D_DRIVER_TYPE_WARP into [D3D12CreateDevice](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-d3d12createdevice). To make development easier, you can use [IDXGIFactory4](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nn-dxgi1_4-idxgifactory4) to deal with both of these. [IDXGIFactory4::EnumAdapterByLuid](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nf-dxgi1_4-idxgifactory4-enumadapterbyluid) (designed to be paired with [ID3D12Device::GetAdapterLuid](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-getadapterluid)) enables an app to retrieve information about the adapter where a Direct3D 12 device was created. [IDXGIFactory4::EnumWarpAdapter](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nf-dxgi1_4-idxgifactory4-enumwarpadapter) provides an adapter which can be provided to D3D12CreateDevice to use the WARP renderer.

_from https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/dxgi-1-4-improvements#cheaper-adapter-enumeration_

We should run some experiments to see what works and what doesn't. I will open an issue to track this.

For this PR, I modified the tests to reuse the instance/adapter and they are now passing.